### PR TITLE
Lock on initialize

### PR DESCRIPTION
**Pull request by hensha256**
## Related Issue
Which issue does this pull request resolve?
Now that we are adding the ability of hooks to Access Lock (#404), we need to prevent external contacts from being able to do anything malicious by calling `initialize` with a hook address that can then access the lock on other functions.

### CODE REVIEW

**Review by zhongeric — APPROVED** [review]
Not sure if there are any additional edge cases created by this but implementation wise this looks good!

File: test/PoolManagerInitialize.t.sol [hunk 1]
```diff
@@ -61,23 +61,23 @@ contract PoolManagerInitializeTest is Test, Deployers, GasSnapshot {
...
         } else if (key0.tickSpacing > manager.MAX_TICK_SPACING()) {
             vm.expectRevert(abi.encodeWithSelector(IPoolManager.TickSpacingTooLarge.selector));
-            manager.initialize(key0, sqrtPriceX96, ZERO_BYTES);
+            initializeRouter.initialize(key0, sqrtPriceX96, ZERO_BYTES);
         } else if (key0.tickSpacing < manager.MIN_TICK_SPACING()) {
             vm.expectRevert(abi.encodeWithSelector(IPoolManager.TickSpacingTooSmall.selector));
-            manager.initialize(key0, sqrtPriceX96, ZERO_BYTES);
-        } else if (key0.currency0 > key0.currency1) {
+            initializeRouter.initialize(key0, sqrtPriceX96, ZERO_BYTES);
+        } else if (key0.currency0 >= key0.currency1) {
```
[the remark this answers is not in the document]
Comment (hensha256) [reply test/PoolManagerInitialize.t.sol:71]: It got combined to have 1 fewer `if` statements. Kind of sad that our fuzz tests didnt catch that this test wasnt updated 😭
Might rename the error message to make it clearer

File: src/test/PoolModifyPositionTest.sol [hunk 2]
```diff
@@ -43,13 +46,13 @@ contract PoolModifyPositionTest is PoolTestBase {
...
-            assert(delta0 > 0 || delta1 > 0);
-            assert(!(delta0 < 0 || delta1 < 0));
+            assertTrue(delta0 > 0 || delta1 > 0, "No positive delta");
+            assertTrue(!(delta0 < 0 || delta1 < 0), "Negative delta on deposit");
             if (delta0 > 0) _settle(data.key.currency0, data.sender, delta.amount0(), true);
             if (delta1 > 0) _settle(data.key.currency1, data.sender, delta.amount1(), true);
         } else {
-            assert(delta0 < 0 || delta1 < 0);
-            assert(!(delta0 > 0 || delta1 > 0));
+            assertTrue(delta0 < 0 || delta1 < 0 || data.key.fee.hasHookWithdrawFee(), "No negative delta");
```
Comment (hensha256) [inline src/test/PoolModifyPositionTest.sol:54]: CI fuzzing randomly caught this edge case that it hasnt caught before 🤷‍♀️

File: .forge-snapshots/initialize.snap [hunk 3]
```diff
@@ -1 +1 @@
-53843
\ No newline at end of file
+74927
```
Comment (snreynolds) [inline .forge-snapshots/initialize.snap:1]: 💀

**Review by snreynolds — APPROVED** [review]

File: .forge-snapshots/initialize.snap (hunk 3 above)
Comment (hensha256) [reply .forge-snapshots/initialize.snap:1]: This does include the asserts in the test - this isnt the complete initialize cost

Comment (snreynolds) [reply .forge-snapshots/initialize.snap:1]: ah true
